### PR TITLE
[3.0] Improve EclipseLink query parameter binding behavior

### PR DESCRIPTION
**Pull request by dazey3**
for #1504

### CODE REVIEW

**Review by lukasj — APPROVED** [review]
LGTM

**Comment by lukasj** [comment]
actually `org.eclipse.persistence.testing.tests.queries.QueryFrameworkTestSuite$4.GetSQLTest` test is failing with this change

**Comment by dazey3** [comment]
@lukasj 

So, I'm a bit confused looking at the purpose of the test https://github.com/eclipse-ee4j/eclipselink/blob/cfda40aec32dc1ffde6f631797195e9ed8dcae80/foundation/eclipselink.core.test/src/it/java/org/eclipse/persistence/testing/tests/queries/QueryFrameworkTestSuite.java#L432-L449

The test seems to be asserting that if the translated query contains "?", then it was translated incorrectly.

Before this change:
```sql
    SELECT ID, FIRSTNAME, LASTNAME FROM employee WHERE (FIRSTNAME = 'Bob')
```

After this change:
```sql
    SELECT ID, FIRSTNAME, LASTNAME FROM employee WHERE (FIRSTNAME = ?)
```

Question:
1) Why does this test expect no parameter marker? The test is calling `org.eclipse.persistence.expressions.Expression.getParameter(String parameterName)` to create a ParameterExpression, but the test expects the translated SQL string to have a literal value? Is that what we should expect? I mean, `org.eclipse.persistence.expressions.Expression.literal(String literal)` exists...

If you agree with the new behavior I am delivering, I can change the test and add one for `org.eclipse.persistence.expressions.Expression.literal(String literal)`, which doesn't exist. If you don't agree with the new behavior, read below.

----------------

Looking at the code changes I am making, I can see why this behavior has changed:

https://github.com/eclipse-ee4j/eclipselink/blob/cfda40aec32dc1ffde6f631797195e9ed8dcae80/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/queries/DatabaseQuery.java#L1499-L1510

The change that I am making is removing the arbitrary `call.setUsesBinding(false);`. 

Question:
1) What is the purpose/usecase for this API call `org.eclipse.persistence.queries.DatabaseQuery.getTranslatedSQLString()`?

If this behavior change is unacceptable, I can put it back the way it is. In my opinion, it seems rather arbitrary to be disabling parameter binding here. But I did review the usage of `getTranslatedSQLString()` and it appears to only be used in testing.

**Comment by dazey3** [comment]
@lukasj @rfelcman 

It seems the build is failing to build DBWS:
```
[INFO] EclipseLink DBWS Builder ........................... FAILURE [  4.017 s]
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.2.0:jar (attach-javadocs) on project org.eclipse.persistence.dbws.builder: MavenReportException: Error while generating Javadoc: 
[ERROR] Exit code: 1 - /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:303: warning: no @param for operations
[ERROR]     public List<CompositeDatabaseType> buildTypesList(List<OperationModel> operations) {
[ERROR]                                        ^
[ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:303: warning: no @return
[ERROR]     public List<CompositeDatabaseType> buildTypesList(List<OperationModel> operations) {
[ERROR]                                        ^
[ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:356: warning: no @param for nct
[ERROR]     public void buildOROXProjects(NamingConventionTransformer nct) {
[ERROR]                 ^
[ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:364: warning: no @param for nct
```

This doesn't seem to be related to my change. Is this a known issue?

**Comment by rfelcman** [comment]
> @lukasj @rfelcman
> 
> It seems the build is failing to build DBWS:
> 
> ```
> [INFO] EclipseLink DBWS Builder ........................... FAILURE [  4.017 s]
> ...
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.2.0:jar (attach-javadocs) on project org.eclipse.persistence.dbws.builder: MavenReportException: Error while generating Javadoc: 
> [ERROR] Exit code: 1 - /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:303: warning: no @param for operations
> [ERROR]     public List<CompositeDatabaseType> buildTypesList(List<OperationModel> operations) {
> [ERROR]                                        ^
> [ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:303: warning: no @return
> [ERROR]     public List<CompositeDatabaseType> buildTypesList(List<OperationModel> operations) {
> [ERROR]                                        ^
> [ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:356: warning: no @param for nct
> [ERROR]     public void buildOROXProjects(NamingConventionTransformer nct) {
> [ERROR]                 ^
> [ERROR] /home/jenkins/agent/workspace/nk-pull-request-verifier_PR-1523/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/BaseDBWSBuilderHelper.java:364: warning: no @param for nct
> ```
> 
> This doesn't seem to be related to my change. Is this a known issue?



Hm Javadoc
`
Error:  /home/runner/work/eclipselink/eclipselink/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/XmlEntityMappingsGenerator.java:642: error: element not closed: ul
Error:       * <ul>
Error:         ^
`
Unclosed `<li>` insided?

File: utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/XmlEntityMappingsGenerator.java [hunk 1]
```diff
@@ -632,14 +630,30 @@ protected static EmbeddedAccessor processEmbeddedMapping(AggregateMapping mappin
...
+    /**
+     * Return a parameter direction as a String based on a given in value.
+     * 
+     * Expected 'direction' value is one of:
+     * <ul>
+     * <li>org.eclipse.persistence.internal.databaseaccess.DatasourceCall.IN
+     * <li>org.eclipse.persistence.internal.databaseaccess.DatasourceCall.INOUT
+     * <li>org.eclipse.persistence.internal.databaseaccess.DatasourceCall.OUT
+     * <li>org.eclipse.persistence.internal.databaseaccess.DatasourceCall.OUT_CURSOR
+     *
```
[the remark this answers is not in the document]
Comment (dazey3) [reply utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/XmlEntityMappingsGenerator.java:647]: ahh, ty for finding that! I'll get it updated

**Comment by dazey3** [comment]
@rfelcman 
> Hm Javadoc
> Error: /home/runner/work/eclipselink/eclipselink/utils/org.eclipse.persistence.dbws.builder/src/main/java/org/eclipse/persistence/tools/dbws/XmlEntityMappingsGenerator.java:642: error: element not closed: ul Error: * <ul> Error: ^
Unclosed <li> insided?

It's written the same way in `master` and I'm not seeing the same failures on `master`. Granted, the missing `</ul>` is missing here on 3.0 and not on `master`, so I fixed that. But if `</li>` is required, why is `master` not failing the same way?

**Comment by dazey3** [comment]
@lukasj @rfelcman 

Looks like the addition of the missing `</ul>` resolved the DBWS failures and it looks good now. I merged the changes back into `2.7` as well.

**Pull request merged**